File: codility/go/6-sorting.go
package golang

import "sort"

func distinct(A []int) int {
	// track number of distinct values in a map
	distinct := make(map[int]struct{})
	for _, num := range A {
		distinct[num] = struct{}{}
	}
	return len(distinct)
}

func maxProductOfThree(A []int) int {
	// largest product of three numbers will always be either:
	// product of three largest numbers (if all positive)
	// product of two smallest numbers and largest number (if two largest numbers are negative)
	sort.Ints(A)
	return max(A[0]*A[1]*A[len(A)-1], A[len(A)-1]*A[len(A)-2]*A[len(A)-3])
}

func triangle(A []int) int {
	// sort the array
	sort.Ints(A)
	// for each value, check if the next two values sum to more than the current value
	for i := 0; i < len(A)-2; i++ {
		if A[i]+A[i+1] > A[i+2] {
			return 1
		}
	}
	return 0
}

func numberOfDiscIntersections(A []int) int {
	counts := make(map[int]*struct{ open, closed int })
	for i, radius := range A {
		if _, ok := counts[i-radius]; !ok {
			counts[i-radius] = &struct{ open, closed int }{}
		}
		counts[i-radius].open++
		if _, ok := counts[i+radius]; !ok {
			counts[i+radius] = &struct{ open, closed int }{}
		}
		counts[i+radius].closed++
	}

	// sort the counts by their pos
	positions := make([]int, 0, len(counts))
	for index := range counts {
		positions = append(positions, index)
	}
	sort.Ints(positions)

	// for each position, check if the number of open discs is greater than the number of closed discs
	open := 0
	intersections := 0
	for _, pos := range positions {
		// open all discs are this position and add them to the number of intersections
		for counts[pos].open > 0 {
			intersections += open
			if intersections > 10000000 {
				return -1
			}
			counts[pos].open--
			open++
		}

		// close all discs at this position
		open -= counts[pos].closed
	}

	return intersections
}

File: codility/py/6-sorting.py
def distinct(A):
    A.sort()
    count = 1
    for i in range(1, len(A)):
        if A[i] != A[i - 1]:
            count += 1

    return count


assert distinct([2, 1, 1, 2, 3, 1]) == 3
assert distinct([4, 6, 3, 5, 6, 4, 5, 5]) == 4

def max_product_of_three(A):
    A.sort()
    return max(
        A[0] * A[1] * A[-1],  # Covers scenario with two negative numbers
        A[-1] * A[-2] * A[-3]  # Covers all others: all pos, all neg, one neg
    )


assert max_product_of_three([-3, 1, 2, -2, 5, 6]) == 60
assert max_product_of_three([-5, 5, -5, 4]) == 125
assert max_product_of_three([-5, -6, -4, -7, -10]) == -120
assert max_product_of_three([-4, -6, 3, 4, 5]) == 120
assert max_product_of_three([-4, -6, 3, 4, 5, 10]) == 240
assert max_product_of_three([-4, -6, 3, 4, 5, 10, 11]) == 550
assert max_product_of_three([-5, -4, -3, -2, 3]) == 60

def triangle(N):
    """
    In a sorted array P < Q < R always is true that P+R>Q and Q+R>P. Only needed checkis for non-obvious case P+Q>R (Find sum of any two consecutive nums bigger than the following one)
    Edge cases: negative nums, len N < 3, in other languages also overflow
    """

    if len(N) < 3:
        return 0

    N.sort()
    # In an example 1 2 5 8 10 20, going from 2 to 10 is enough
    for i in range(1,len(N)-1):
        if (N[i]+N[i-1]) > N[i+1]:
            return 1
    return 0


assert triangle([10, 2, 5, 1, 8, 20]) == 1
assert triangle([10, 50, 5, 1]) == 0
assert triangle([]) == 0
assert triangle([1, 2]) == 0
assert triangle([5, 5, 5]) == 1
assert triangle([-5, -3, -1, 2, 8]) == 0
assert triangle([2147483647, 2147483647, 2147483647]) == 1
